### PR TITLE
Windows support and Ipopt 3.12.13

### LINKED ISSUE #1: Please support Windows

### PR DESCRIPTION
**Pull request by richardotis**
Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Closes #1 

This PR is an attempt to get Windows support working for ipopt using the m2w64 toolchain and cmake. The cmake scripts are based on the awesome efforts of @ycollet to ship CMake scripts for the whole Coin-OR ecosystem: https://github.com/ycollet/coinor-cmake
The scripts here are lightly modified to address some path differences, and for stability across future versions of Ipopt.

1. Do we need to switch all three platforms to CMake here, or is it okay to only use CMake for Windows?
2. I'm using m2w64-toolchain here, even though it's deprecated, because the `compiler('m2w64_c')` syntax didn't work locally when I ran `conda build recipe -m .ci_support\win_c_compilervs2015cxx_compilervs2015.yaml`. Help would be appreciated here.
3. The CMake scripts from @ycollet are LGPL licensed (for which I don't really understand the implications in the context of build scripts). I think that accepting this PR would mean this feedstock (but not the actual package) would become LGPL licensed. Someone from the @conda-forge/core  team should probably chime in here.
4. There are likely a bunch of unnecessary CMake support macros here that can be deleted.
5. The requirements in the recipe are not fully optimized, and I am not 100% clear on the differences between build, host and run requirements (I read the documentation and still don't quite get it).
6. Ipopt has a bunch of build-time options, including for AMPL, HSL, Metis, performance options like fastmath and MKL, as well as minor(?) ones like link-time optimization. Which of these should be exposed as features in the recipe?
7. There is a lot of path- and link-hacking in the bld.bat to get the compiler to find and link all the needed libraries. I am not sure how portable it is across Windows version, nor am I clear on why CMake is not doing this automatically (it's there in the CMake scripts).

### CODE REVIEW

**Comment by chrisdembia** [comment]
Neat! Where can I learn more about the `m2w64` toolchain? Which compilers are used? Will this build of Ipopt be usable with Visual Studio?

**Comment by richardotis** [comment]
@chrisdembia I don't know the best place; everything I've learned about m2w64 has been random conda-forge feedstock issue trackers. There is some discussion here: https://github.com/ContinuumIO/anaconda-issues/issues/1484 The toolchain package itself is deprecated but the technology has been rolled into conda-forge's official compiler support, so I think it's still okay to use (but it has a new name). 

MSVC does not appear to compile Ipopt (I tried that first), and it is not officially supported upstream. The compiler stack used here is essentially MinGW.

**Comment by chrisdembia** [comment]
Okay thank you. I'm hoping to use Ipopt in a C++ project compiled with Visual Studio, and my understanding is I cannot use libraries compiled with MinGW.

I think Ipopt can be compiled with Visual Studio; it's just that its MUMPS dependency is written in Fortran and needs to be compiled with gfortran or another open-source Fortran compiler.

Anyway, I was just curious. You can ignore my comments.

**Comment by richardotis** [comment]
@conda-forge-admin, please rerender

**Comment by richardotis** [comment]
@conda-forge/core It looks to me like something weird is happening in the Linux build infrastructure. Configure seems to be pulling in the link flag `-lm'` (note the extra quote character). Is that normal? If not, it would explain the strange `linking to Fortran libraries from C fails` message we're getting here.

**Comment by richardotis** [comment]
I can reproduce the Linux error locally with ` conda build recipe -m .ci_support/linux_.yaml`

**Comment by richardotis** [comment]
Here's an excerpt from the config.log:
```
configure:2027: $? = 0
configure:2029: /home/rotis/anaconda3/conda-bld/ipopt_1577132049035/_build_env/bin/x86_64-conda_cos6-linux-gnu-cc -v </dev/null >&5
Reading specs from /home/rotis/anaconda3/conda-bld/ipopt_1577132049035/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/specs
COLLECT_GCC=/home/rotis/anaconda3/conda-bld/ipopt_1577132049035/_build_env/bin/x86_64-conda_cos6-linux-gnu-cc
COLLECT_LTO_WRAPPER=/home/rotis/anaconda3/conda-bld/ipopt_1577132049035/_build_env/bin/../libexec/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/lto-wrapper
Target: x86_64-conda_cos6-linux-gnu
Configured with: /home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/configure --build=x86_64-build_pc-linux-gnu --host=x86_64-build_pc-linux-gnu --target=x86_64-conda_cos6-linux-gnu --prefix=/home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/gcc_built --with-sysroot=/home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/gcc_built/x86_64-conda_cos6-linux-gnu/sysroot --enable-languages=c,c++,fortran,objc,obj-c++ --with-pkgversion='crosstool-NG 1.23.0.452-d158' --enable-__cxa_atexit --disable-libmudflap --enable-libgomp --disable-libssp --enable-libquadmath --enable-libquadmath-support --enable-libsanitizer --enable-libmpx --with-gmp=/home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/.build/x86_64-conda_cos6-linux-gnu/buildtools --with-mpfr=/home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/.build/x86_64-conda_cos6-linux-gnu/buildtools --with-mpc=/home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/.build/x86_64-conda_cos6-linux-gnu/buildtools --with-isl=/home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/.build/x86_64-conda_cos6-linux-gnu/buildtools --enable-lto --with-host-libstdcxx='-static-libgcc -Wl,-Bstatic,-lstdc++,-Bdynamic -lm' --enable-threads=posix --enable-target-optspace --enable-plugin --enable-gold --disable-nls --disable-multilib --with-local-prefix=/home/conda/feedstock_root/build_artifacts/ctng-compilers_1570984655235/work/gcc_built/x86_64-conda_cos6-linux-gnu/sysroot --enable-long-long --enable-default-pie
Thread model: posix
gcc version 7.3.0 (crosstool-NG 1.23.0.452-d158) 
```

Notice we're getting back (emphasis mine) `--with-host-libstdcxx='-static-libgcc -Wl,-Bstatic,-lstdc++,-Bdynamic **-lm'** `

This looks like a bad interaction between the gcc output and Ipopt's configure script. This is a known upstream bug for a long time: https://projects.coin-or.org/Ipopt/ticket/215 - (So how did the Linux build ever work...?)

**Comment by richardotis** [comment]
This workaround commit was just a test to see if the Linux build will work if you hardcode FLIBS. The real solution (which should probably become its own PR) will be to patch the upstream `BuildTools/coin.m4` to pass a correctly formed FLIBS. We're on the hook to do this because the bug is marked wontfix upstream, and the next version of Ipopt actually has a build system overhaul, so any upstream patch would probably be moot.

I'd like to get functioning Ipopt 3.12.x for all platforms before dealing with 3.13, though.

**Comment by richardotis** [comment]
@conda-forge-admin, please rerender

**Comment by richardotis** [comment]
This is ready for review.

**Comment by moorepants** [comment]
Wow, what an effort. Thanks @richardotis !

I personally wouldn't know where to being on reviewing this. My only questions are:

- Is this mingw approach used in other conda-forge packages and do the primary conda-forge folks approve of this method?
- Do ipopt tests pass when this package is installed?
- Is everything installed in some standard locations and proper env variables set so that downstream packages that depend on this will be able to load everything on windows?
- Will downstream packages have to deal with any of the mingw toolchain and/or are there build/install consequences for those packages?

**Comment by richardotis** [comment]
1. I think someone who knows the core infrastructure better than me will have to answer this.
2. The current PR runs some tests at the tail end of the build process (same as for Linux/OSX and was done previously).
3. The DLL and static archive are going into standard locations for the MinGW toolchain, but my experience is that downstream packages have to explicitly add the MinGW library directories; Anaconda/conda-forge doesn't do it for you.
4. I have a small local patch for the cyipopt feedstock which patches setup.py to find the ipopt DLL at its new location. Since the name and location of the DLL will be conda-forge-specific, I figured it's okay for the change to live in the feedstock, as well as the feedstocks for any other downstream packages.

(As an aside, probably for discussion in the cyipopt feedstock PR, cyipopt for Windows currently finds the ipopt DLL, and then copies it into the same directory as the Cython extension for cyipopt. My patch won't change that behavior, but we either need to enable some kind of pinning to ensure cyipopt is rebuilt any time ipopt is, or figure out how to get the ipopt DLL in the MinGW library directory detected at runtime.)

**Comment by moorepants** [comment]
All sounds good. I'd like to hear the answers to the questions at the very top too. We'd need some conda forge people to comment on that.

**Comment by moorepants** [comment]
@conda-forge/staged-recipes Would a conda-forge developer mind having a look at this PR? We want to make sure this mingw method is good to go?

**Comment by moorepants** [comment]
@isuruf Would you have any time to comment on the questions posed at the top of this PR? None of us here are familiar with the best practices in conda-forge regarding mingw builds.

**Comment by moorepants** [comment]
> The CMake scripts from @ycollet are LGPL licensed (for which I don't really understand the implications in the context of build scripts). I think that accepting this PR would mean this feedstock (but not the actual package) would become LGPL licensed. Someone from the @conda-forge/core team should probably chime in here.

I'm not sure if we are supposed to license the build scripts anything other than the default. But I am sure that other feedstocks likely contain patches that are taken from non-BSD-3 licensed code.

File: recipe/build.sh [hunk 1]
```diff
@@ -8,6 +8,7 @@ mkdir build
 cd build
 
 ../configure \
+  FLIBS="-L$PREFIX/lib -L$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0 -L$BUILD_PREFIX/bin/../lib/gcc -L$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/lib/../lib -L$BUILD_PREFIX/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/../lib -L$BUILD_PREFIX/bin/../x86_64-conda_cos6-linux-gnu/sysroot/usr/lib/../lib -L$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/lib -L$BUILD_PREFIX/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib -L$BUILD_PREFIX/bin/../x86_64-conda_cos6-linux-gnu/sysroot/usr/lib -lrt -lgfortran -lm -lquadmath" \
```
Comment (moorepants) [inline recipe/build.sh]: Is this change necessary? Seems like these should be discovered rather than specified explicitly.

Comment (richardotis) [reply recipe/build.sh]: That shouldn't be here and looks like a rebase error on my part.

**Comment by moorepants** [comment]
@isuruf Would you be willing to answer our questions here and double check this windows build?

**Comment by isuruf** [comment]
> Is this mingw approach used in other conda-forge packages and do the primary conda-forge folks approve of this method?

Only rarely. Most packages use MSVC.

> Will downstream packages have to deal with any of the mingw toolchain and/or are there build/install consequences for those packages?

The C runtime that MinGW and MSVC are using are slightly different. Therefore, there might be incompatibilities. (segfaults when you try to free an object created in a different C runtime)
So, downstream packages will have to be built with MinGW too.

**Comment by moorepants** [comment]
Thanks! There is also the list of questions in the very first comment that we don't have answers to.

**Comment by isuruf** [comment]
Also, the C++ runtimes are completely different.

**Comment by isuruf** [comment]
1. No. Using CMake just for Windows is fine and preferred to keep backwards compatibility.
2. You need to rerender after using the `compiler('m2w64_c')` syntax
3. That's fine. You can keep the license of the scripts in the recipe.
7. Those are probably bugs in the cmake scripts.

**Comment by isuruf** [comment]
Since there's CMake support, why not try VS2015?

**Comment by richardotis** [comment]
@isuruf One part of the ipopt build process requires a working Fortran compiler, at least for the 3.12.x versions. Also, when I naively tried changing the build script to create CMake VS Makefiles, I received errors unrelated to a missing Fortran compiler. Generally speaking, upstream seemed to recommend the MSYS/MinGW option for building: https://projects.coin-or.org/Ipopt/browser/stable/3.12/Ipopt/doc/documentation.pdf?format=raw The story might have changed for 3.13, since they rewrote the build system, but I haven't looked at it yet.

I'm not saying it's impossible, but it's unclear how much work it would be.

**Comment by moorepants** [comment]
We are are using flang in th slycot feedstock with cmake:

https://github.com/conda-forge/slycot-feedstock

Maybe that's possible?

**Comment by richardotis** [comment]
These new commits for a MSVC-compatible build depend on https://github.com/conda-forge/mumps-feedstock/pull/56

Even with that PR, locally I get some test failures for the built-in examples. However, the majority of them pass and I am not sure if some numerical instability is to be expected here.

**Comment by nniclausse** [comment]
Hi, i tested the 3.12.13 package from the pycalphad channel, it does not include the file IpoptConfig.h which is needed by IpDebug.hpp, IpJournalist.hpp and IpTypes.hpp

**Comment by richardotis** [comment]
@nniclausse Are those files needed by downstream packages, or only at build time?

**Comment by nniclausse** [comment]
> @nniclausse Are those files needed by downstream packages, or only at build time?

I'm trying to build an app that use ipopt, this app includes IpTNLP.hpp, which requires IpUtils.hpp  -> IpTypes.hpp -> IpoptConfig.h -> config_ipopt_default.h

**Comment by richardotis** [comment]
Ah ok, I think the reason I didn't catch it is because I was testing against the C interface without making sure C++ still worked.

**Comment by traversaro** [comment]
Hi @richardotis, I am interesting in helping to get this merged, any suggestion on what are the next steps that need to be addressed? Thanks!

**Comment by traversaro** [comment]
I guess we can close this PR as it has been superseded by https://github.com/conda-forge/ipopt-feedstock/pull/47  @conda-forge/ipopt .
